text/0000-assoc-default-groups.md

```diff
@@ -0,0 +1,1468 @@
+}
+```
+
+In the current implementation, (6) is rejected because the compiler will not
+let you assume that `x` is of type `usize`. But in this proposal, you would be
+allowed to assume this. To permit this is not a problem because `Foo for Vec<T>`
+is not further specializable since `baz` in the implementation has not been
+marked as `default`.
+
+With these changes,
+we consider the design of associated type defaults to be *finalized*.
+
+## `default` specialization groups
+[default_groups]: #default-specialization-groups
+
+Note: Overlapping implementations, where one is more specific than the other,
+requires actual support for [specialization].
+
+Now, you might be thinking: - *"Well, what if I __do__ need to assume that
+my defaulted associated type is what I said in a provided method,
+what do I do then?"*. Don't worry; We've got you covered.
+
+To be able to assume that `Self::Bar` is truly `u8` in snippets (2) and (5),
+you may henceforth use `default { .. }` to group associated items into atomic
+units of specialization. This means that if one item in `default { .. }` is
+overridden in an implementation, then all all the items must be. An example (7):
+
+```rust
+struct Country(&'static str);
+
+struct LangSec { papers: usize }
+struct CategoryTheory { papers: usize }
+
+trait ComputerScientist {
+    default {
+        type Details = Country;
+        const THE_DETAILS: Self::Details = Country("Scotland"); // OK!
+        fn papers(details: Self::Details) -> u8 { 19 } // OK!
+    }
+}
+
+// https://en.wikipedia.org/wiki/Emily_Riehl
+struct EmilyRiehl;
+
+// https://www.cis.upenn.edu/~sweirich/
+struct StephanieWeirich;
+
+// http://www.cse.chalmers.se/~andrei/
+struct AndreiSabelfeld;
+
+// https://en.wikipedia.org/wiki/Conor_McBride
+struct ConorMcBride;
+
+impl ComputerScientist for EmilyRiehl {
+    type Details = CategoryTheory;
+
+    // ERROR! You must override THE_DETAILS and papers.
+}
+
+impl ComputerScientist for StephanieWeirich {
+    const THE_DETAILS: Country = Country("USA");
+    fn papers(details: Self::Details) -> u8 { 86 }
+
+    // ERROR! You must override Details.
+}
+
+impl ComputerScientist for AndreiSabelfeld {    
+    type Details = LangSec;
+    const THE_DETAILS: Self::Details = LangSec { papers: 90 };
+    fn papers(details: Self::Details) -> u8 { details.papers }
+
+    // OK! We have overridden all items in the group.
+}
+
+impl ComputerScientist for ConorMcBride {
+    // OK! We have not overridden anything in the group.
+}
+```
+
+You may also use `default { .. }` in implementations.
+When you do so, everything in the group is automatically overridable.
+For any items outside the group, you may assume their signatures,
+but not the default definitions given. An example:
+
+```rust
+use std::marker::PhantomData;
+
+trait Fruit {
+    type Details;
+    fn foo();
+    fn bar();
+    fn baz();
+}
+
+struct Citrus<S> { species: PhantomData<S> }
+struct Orange<V> { variety: PhantomData<V> }
```

Comment:
nitpick: it looks like everything here would still compile with `struct Citrus<S> { species: S }`, so I'd encourage you to just do that.  `PhantomData` is a subtle and attention-catching enough thing that it probably shouldn't be mentioned in a guide-level explanation unless it's _necessary_ to the feature.

Reply:
Makes sense :) I'll fix later today.